### PR TITLE
fix: numexpr to fix CVE-2023-39631⁠ (2.8.4 => 2.9.0)

### PR DESCRIPTION
**Pull request by nigzak**
### SUMMARY
superset 3.1.0 / 3.1.1 has a critical finding: CVE-2023-39631⁠ because of numexpr 2.8.4

### TESTING INSTRUCTIONS
unknown (sorry), this pull request is to update the numexpr to a non-faulty version depending the CVE (2.9.0 is the current latest available version based on https://github.com/pydata/numexpr)
To test "my" finding: open final image in docker scout => no finding for numexpr is there anymore

image

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/26967
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

HINT: This is my first pull request ... if I made something wrong please let me know that I can make a next one better :)

### CODE REVIEW

**Comment by nigzak** [comment]
Hint: made a "manual dirty" update in the docker image to update to 2.9.0 (which for sure kills starting the docker image).

```
from apache/superset:3.1.1

USER root:root

RUN pip uninstall numexpr --yes
RUN pip install numexpr==2.9.0


USER superset:superset
```

With this update there is no critical finding of the CVE-2023-39631⁠ in the container anymore - an update of numexp should finally fix this finding

image

**Comment by nigzak** [comment]
fixed name of PR (seems I forgot the ":" after chore)

**Comment by john-bodley** [comment]
@nigzak it seems like `numexpr` is an indirect dependency:

```
$ git grep numexpr
requirements/base.txt:numexpr==2.8.4
requirements/base.txt:    #   numexpr
```

To ensure that CVE is addressed for all situations (and not just CI etc. which leverage the pinned/frozen dependencies) we should add the necessary bump to `setup.py`. This should _likely_ be done by updating the `pandas` constraints given that's the package which is depending on it.

**Comment by nigzak** [comment]
HI @dpgaspar 
that is bad to me, I am not allowed to deploy this version because it has a critical CVE inside ... so I must stay to superset 3.0.x ... (this has not included this dependency)

I even don't know what others do ... but at least normally - as I know - deploying software which having critical or even high findings is not allowed - or only with "hard discussions" (more or less mostly only MAX for high or where no fix is avaiable).

This "panda" thing I did not know yet so I sadly cannot discuss anything about this ... I tought it is a "normal" package like usual in requirements for pip

As I understood as result this pull request then will be rejected and you will stay on the bad version - right?

**Comment by nigzak** [comment]
made a pull request on panda site to update to 2.9.0 as well - lets see how they take it up :)

**Comment by dpgaspar** [comment]
> made a pull request on panda site to update to 2.9.0 as well - lets see how they take it up :)

I think it is acceptable/desirable to update the `requirement/base.in` and `base.txt` like your already doing (don't forget to run `pip-compile-multi --no-upgrade` to check everything and update the comment with references

File: requirements/development.txt [hunk 1]
```diff
@@ -6,12 +6,10 @@
 #    pip-compile-multi
 #
 -r base.txt
--e file:.
+-e file:///mnt/c/temp/superset_numexp
```
Comment (dpgaspar) [inline requirements/development.txt:9]: remove this and replace it with `.` like it was before

File: requirements/docker.txt [hunk 2]
```diff
@@ -6,7 +6,7 @@
 #    pip-compile-multi
 #
 -r base.txt
--e file:.
+-e file:///mnt/c/temp/superset_numexp
```
Comment (dpgaspar) [inline requirements/docker.txt:9]: remove this and replace it with `.` like it was before

File: requirements/local.txt [hunk 3]
```diff
@@ -6,7 +6,7 @@
 #    pip-compile-multi
 #
 -r development.txt
--e file:.
+-e file:///mnt/c/temp/superset_numexp
```
Comment (dpgaspar) [inline requirements/local.txt:9]: remove this and replace it with `.` like it was before

File: requirements/testing.txt [hunk 4]
```diff
@@ -7,7 +7,7 @@
 #
 -r development.txt
 -r integration.txt
--e file:.
+-e file:///mnt/c/temp/superset_numexp
```
Comment (dpgaspar) [inline requirements/testing.txt:10]: remove this and replace it with `.` like it was before

File: requirements/base.txt [hunk 5]
```diff
@@ -90,6 +90,8 @@ dnspython==2.1.0
     # via email-validator
 email-validator==1.1.3
     # via flask-appbuilder
+exceptiongroup==1.2.0
+    # via cattrs
```
Comment (dpgaspar) [inline requirements/base.txt:94]: I've executed `pip-compile-multi --no-upgrade` and got no such changes, are you based on the latest master on this PR? anyway I think these should be removed

**Review by dpgaspar — COMMENTED** [review]
please add `numexpr>=2.9.0` to `base.in`

File: requirements/development.txt (hunk 1 above)
Comment (nigzak) [reply requirements/development.txt]: fixed it back to .

File: requirements/docker.txt (hunk 2 above)
Comment (nigzak) [reply requirements/docker.txt]: fixed it back to .

File: requirements/local.txt (hunk 3 above)
Comment (nigzak) [reply requirements/local.txt]: fixed it back to .

File: requirements/testing.txt (hunk 4 above)
Comment (nigzak) [reply requirements/testing.txt]: fixed it back to .

File: requirements/base.txt (hunk 5 above)
Comment (nigzak) [reply requirements/base.txt:94]: I don't know where this comes from, I did not change something else?
I made a re-sync with the master and run it again, it still appears in my output?

If I am looking to the package "cattrs" it seems it is a mandatory requirement and it should be there?
I can also manual remove it if it is not good?

https://github.com/apache/superset/blob/master/requirements/base.txt
```
cattrs==23.2.1
    # via requests-cache
```

if I check this package local it says that exceptiongroup is a mandatory sub-dependency?
```
....pip show cattrs
Name: cattrs
Version: 23.2.1
Summary: Composable complex class support for attrs and dataclasses.
Home-page:
Author:
Author-email: Tin Tvrtkovic <tinchester@gmail.com>
License: MIT
Location: /home/starnol/.local/lib/python3.10/site-packages
Requires: attrs, exceptiongroup, typing-extensions
Required-by:
```

shall i still remove it?

Comment (nigzak) [reply requirements/base.txt:94]: I am using the Ubuntu subsystem (WSL2) in windows with the normal installed python which you can get with apt-get

```
... $ python3 --version
Python 3.10.12

.... ~$ pip --version
pip 24.0 from /home/..../.local/lib/python3.10/site-packages/pip (python 3.10)

// no updates available
~$ sudo apt-get upgrade -y
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages have been kept back:
  ubuntu-advantage-tools
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.

// no update for python3
~$ sudo apt-get install python3
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
python3 is already the newest version (3.10.6-1~22.04).
python3 set to manually installed.
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
```

Comment (dpgaspar) [reply requirements/base.txt:94]: Thank you, I thinks this is correct

**Comment by dpgaspar** [comment]
> please add `numexpr>=2.9.0` to `base.in`

Add this ^^^ and run pip-compile-multi again, please

**Comment by nigzak** [comment]
Done @dpgaspar 
... sorry, I have read over this detail

**Comment by rusackas** [comment]
Running CI... 🤞

**Comment by dpgaspar** [comment]
@nigzak can you please rebase and fix conflicts

**Comment by nigzak** [comment]
I am not sure if I made the rebase correct ... sorry ... this is not my daily thing to do, I tried my best to get it to work ... :/
At least I got a confirmation in the bash

```
...
Successfully rebased and updated refs/heads/master.
```
also the push with force afterwards was sucessfully. 
I think it should be right? If not please let me know ... again sorry, I still try to learn how this all works ...

**Comment by dpgaspar** [comment]
hey @nigzak I'm sorry but you have to rebase again or merge an updated master into your branch, this is to fix docker-build step on CI

**Comment by nigzak** [comment]
rebase local failed (added the previous commits to my commit in addition) so I re-synced the forc with the master and commited again the changes to avoid this. Hope now it is fine?

**Comment by rusackas** [comment]
Looks like CI is happy! Maybe @dpgaspar can give it another look before it needs another rebase 😅

**Comment by nigzak** [comment]
synced fork again with master, executed "pip-compile-multi --no-upgrade" and pushed again to git

**Review by dpgaspar — APPROVED** [review]

**Pull request merged**